opm/autodiff/VFPHelpers.hpp

```diff
@@ -764,8 +764,93 @@ inline double findTHP(
 
 
 
+// a data type use to do the intersection calculation to get the intial bhp under THP control
+struct RateBhpPair {
+    double rate;
+    double bhp;
+};
+
+
+// looking for a intersection point a line segment and a line, they are both defined with two points
+// it is copied from #include <opm/polymer/Point2D.hpp>, which should be removed since it is only required by the lagacy polymer
+inline bool findIntersection(const std::array<RateBhpPair, 2>& line_segment, const std::array<RateBhpPair, 2>& line, double& bhp) {
+    const double x1 = line_segment[0].rate;
+    const double y1 = line_segment[0].bhp;
+    const double x2 = line_segment[1].rate;
+    const double y2 = line_segment[1].bhp;
+
+    const double x3 = line[0].rate;
+    const double y3 = line[0].bhp;
+    const double x4 = line[1].rate;
+    const double y4 = line[1].bhp;
+
+    const double d = (x1 - x2) * (y3 - y4) - (y1 - y2) * (x3 - x4);
+
+    if (d == 0.) {
+        return false;
+    }
+
+    const double x = ((x3 - x4) * (x1 * y2 - y1 * x2) - (x1 - x2) * (x3 * y4 - y3 * x4)) / d;
+    const double y = ((y3 - y4) * (x1 * y2 - y1 * x2) - (y1 - y2) * (x3 * y4 - y3 * x4)) / d;
+
+    if (x >= std::min(x1,x2) && x <= std::max(x1,x2)) {
+        bhp = y;
+        return true;
+    } else {
+        return false;
+    }
+}
 
+// calculating the BHP from thp through the intersection of VFP curves and inflow performance relationship
+inline bool findIntersectionForBhp(const std::vector<RateBhpPair>& ratebhp_samples,
+                                   const std::array<RateBhpPair, 2>& ratebhp_twopoints_ipr,
+                                   double& obtained_bhp)
+{
+    // there possibly two intersection point, then we choose the one corresponding with the bigger rate
+
```

Comment:
What is the rationale for this choice?

Reply:
following manual. although this is the only information manual gives in a clear way. A lot of other choices by us are in other places.